### ISSUE TITLE
Automatic class_weights/scale_pos_weight

### ISSUE THREAD

**Issue by aparamon** [catboost/catboost#392]
Currently, one can specify `class_weights` or `scale_pos_weight` parameter to help CatBoost treat imbalanced training sets. A sensible value for `scale_pos_weight` is `num_neg`/`num_pos`; it is proposed that CatBoost gets additional parameter `auto_class_weights` which would calculate and set `class_weights` based on actual `num_neg`, `num_pos` of class records in the training set.

**Comment by thomas9000** [comment]
Any update on this issue? I find it troublesome to hardcode the class_weights parameter manually each time I change my training datasets with a different num_neg/num_pos.

**Comment by Sandy4321** [comment]
yes it would be great to implement this

**Comment by ohld** [comment]
Any update on this feature?

**Comment by annaveronika** [comment]
Still waiting for contributors. I think it's a very doable issue, you could try implementing this one!

**Comment by nlauchande** [comment]
@annaveronika: I am interested in trying to get started in contributing to Catboost. Can you guide me on where to look to make this change?

**Comment by ShaharKatz** [comment]
We can use this study:
[Link to original paper](https://pdfs.semanticscholar.org/8939/585e7d464703fe0ec8ca9fc6acc3528ce601.pdf) 
Attaching the paper annotated with my notes - [The Effect of Class Distribution on Classifier Learning- An Empirical Study 2001_annotated.pdf](https://github.com/catboost/catboost/files/4808593/The.Effect.of.Class.Distribution.on.Classifier.Learning-.An.Empirical.Study.2001_annotated.pdf)

If this hasn't been picked up i'll try adding it

**Comment by Sandy4321** [comment]
Shahar, great paper  and great idea
Just do it
We do need this, since every real data is unbalanced

**Comment by Sandy4321** [comment]
Shahar 
https://github.com/ShaharKatz
also you can contribute to this 
https://github.com/scikit-learn-contrib/imbalanced-learn
https://github.com/scikit-learn-contrib/imbalanced-learn/issues/730

**Comment by ShaharKatz** [comment]
Thanks, saw the open issue there (which is probably a more classic case).
Forked catboost, hopefully most of the code should be the same for both repos

**Comment by LyzhinIvan** [comment]
Hi, @ShaharKatz ! I have read the paper and understood the method for simple decision tree. But I can't understand how we should reassign leaf values in case of gradient boosting? Do you have formulas for this?

**Comment by ShaharKatz** [comment]
No, for gradient boosting we will have to do the fix per tree. I'm currently working on solving this in the simple scikit-learn case and if this proves valuable i'll also use this here. I think it's a good idea to first test the method in an easier setup and see that it (1) works and (2) provides value

**Comment by LyzhinIvan** [comment]
Ok, then I close this issue now, because auto-class-weights option is added already. Feel free to reopen the issue when you will be ready to contribute.

**Comment by Sandy4321** [comment]
, then I close this issue now, because auto-class-weights option is added already
Can you share full code example?
